neps/nep-504.md

```diff
@@ -0,0 +1,249 @@
+
+## Motivation
+
+[NEP-393](https://github.com/near/NEPs/pull/393) defines a standard for Soul Bound Tokens, which are a foundational for Decentralized Society. SBTs can represent certificates, badges and anything that should be bound to a "soul".
+The contract issuers often need more details than what NEP-393 defines. Such details are meant to be set in the `reference` field, which has an open, non standardized format.
+
+NEAR doesn't have a well structured Metadata standard. The [NFT Metadata](https://github.com/near/NEPs/blob/master/neps/nep-0177.md) doesn't clearly define the `reference` extension. In fact, the NEP-393 Metadata is defined based on NFT Metadata. The schema defined here can be used by other token variations as well.
+
+Issuers, contract implementers and users demand a guideline and a minimum schema to make better integration with tools and projects. This document defines a minimum common schema for reference fields of `ContractMetadata` and `TokenMetadata`. Moreover, while working with various issuers, we found that issuers and minters often duplicate lot of data in the `TokenMetadata`. We found that it is unnecessary and it would be useful to create a new type: `ClassMetadata`, that defines common data for all tokens of a given (issuer, class) pair.
+
+## Specification
+
+NEP-393 `TokenMetadata` is minimalistic. It only requires data which a contract can reason about, and all user facing details (such as title, description) are meant to be part of the reference.
+Similarly, `ContractMetadata` only defines issuer basic info.
+
+### Class Metadata
+
+We start by defining a `ClassMetadata` which will contain all common information about tokens of the same class. The `ClassMeetadata` object must be stored in the SBT issuer contract and accessible via `class_metadata(class: ClassId)` query.
+This will allow to avoid duplicating such information in all `TokenMetadata` and potentially allow to leave `TokenMetadata.reference` empty, if the tokens only differentiate by the class, issued_at and expires_at attributes.
+
+```rust
+pub struct ClassMetadata {
+    /// Issuer class name. Required.
+    pub name: String,
+    /// If defined, should be used instead of `contract_metadata.symbol`.
+    pub symbol: Option<String>,
+    /// An URL to an Icon. To protect fellow developers from unintentionally triggering any
+    /// SSRF vulnerabilities with URL parsers, we don't allow to set an image bytes here.
+    /// If it doesn't start with a scheme (eg: https://)
+    /// then `contract_metadata.base_uri` should be prepended.
+    pub icon: Option<String>,
+    /// JSON or an URL to a JSON file with more info. If it doesn't start with a scheme
+    /// (eg: https://) then base_uri should be prepended.
+    pub reference: Option<String>,
+    /// Base64-encoded sha256 hash of JSON from reference field. Required if `reference` is included.
+    pub reference_hash: Option<Base64VecU8>,
```

Comment:
@robert-zaremba Let me try to deliver my point once again:

https://github.com/near/NEPs/blob/6eee9dc14ff64dd2eac24e4ae704e0b8fbb18ba9/neps/nep-0177.md#L45-L46

`reference` field is designed to be a **link** and nothing else.

If you want to have more fields stored as part of `ClassMetadata`, just add those fields on the same level as `name`, `symbol`, ... fields - this is not a breaking change and won't affect consumers of the API as they will just ignore unsupported fields. There is no need in string-encoding JSON data into the `reference` field.

Reply:
> My reasoning is to have a strict minimum amount of keys in ClassMetadata, and add more structure for other common requirements, so tools and websites can display more information about tokens.

There is no reason to overcomplicate things. If you want to standardize fields, add them to ClassMetadata optional or not. `reference` is needed when the metadata is too big to be stored inline. There is no reason to store JSON as a string inside JSON value.

> In such case, we would need to extend the ClassMetadata as more requirements will emerge from the ecosystem. So far, for the SBTs we were working with, we need different pictures (size / format), website address, social_media links ...

If you will want to standardize on those, you will still need to extend the NEP and some structure, be it `ClassMetadata` or some other struct, it does not really change things. Also, you don't always need to standardize every single field as that may limit the usability of the whole spec if those fields will be used sparsely.

> I would like to hear more opinions from the tool designers and other projects dealing with tokens.

Sure, make a call in community channels.

Reply:
> you will still need to extend the NEP and some structure

This is what an appendix is for.

We need to have a structure to make sure tools will have a common way to handle the data we describe in this PR. Can be in reference (as originally proposed), or by adding more optional attributes to the top level Metadata. Either way we need to have that structure.

Reply:
> We need to have a structure to make sure tools will have a common way to handle the data we describe in this PR. Can be in reference (as originally proposed), or by adding more optional attributes to the top level Metadata. Either way we need to have that structure.

I don't argue against having the defined structure, I just suggest moving it to where it belongs (new optional fields on `ClassMetadata`), instead of abusing `reference` field.

Reply:
Sure, I'm OK with that, see the second part of the [post](https://github.com/near/NEPs/pull/504#discussion_r1376331389).

Reply:
> ClassMetadata is not open ended.

It is, actually, open ended by JSON spec, and that should be encouraged to be used whenever it makes sense.

> Sure, I'm OK with that, see the second part of the https://github.com/near/NEPs/pull/504#discussion_r1376331389.

Alright, so you are waiting for inputs from someone else besides me, right?